### PR TITLE
OOPs session 1 Assignments

### PR DESCRIPTION
**Pull request by meta-udit-saxena**

### CODE REVIEW

File: Calculator/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+public class Calculator {
+
+	/**
+	 * perform controlling function of taking input pass it for evaluating
+	 * display the result
+	 * 
+	 */
+	public void actionPerformed() {
```
Comment (chirag-metacube) [inline Calculator/Calculator.java:15]: actionPerformed stating that some action is already performed, this can be performAction instead

File: Calculator/Display.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/*
+ * display the application and use layouts from Layout class  
+ */
+public class Display extends Layout {
+	/**
+	 * shows final result displayed on screen
+	 * 
+	 * @param result
+	 */
+	public void show(int result) {
```
Comment (chirag-metacube) [inline Calculator/Display.java:10]: Calculator result should be decimal instead of integer

File: Calculator/Layout.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import java.awt.Component;
+
+/**
+ * determine the layout
+ * @author Udit Saxena
+ *
+ */
+public class Layout {
+
+	private int screenSize[][];
```
Comment (chirag-metacube) [inline Calculator/Layout.java:10]: Screen size can be in just 2 variables: Height and width, No need to take complete 2D array

File: Calculator/Layout.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	 * @param comp
+	 *            - Component Class object
+	 */
+	void addLayoutComponent(Component comp) {
+	}
+
+	/**
+	 * Calculates the minimum size dimensions, given the components it contain
+	 */
+	void minimumLayoutSize() {
```
Comment (chirag-metacube) [inline Calculator/Layout.java:25]: Method name should specify whether it is used to set or get. This can be setMinimumLayoutSize. and also some size needs to be passed in this method as parameters.

File: Calculator/Layout.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	private int screenSize[][];
+
+	/**
+	 * Adds the component comp to the layout, associating it with the string
+	 * specified by name
+	 * 
+	 * @param comp
+	 *            - Component Class object
+	 */
+	void addLayoutComponent(Component comp) {
```
Comment (chirag-metacube) [inline Calculator/Layout.java:19]: Missing access modifier

File: Calculator/Layout.java (hunk 4 above)
Comment (chirag-metacube) [reply Calculator/Layout.java:25]: Missing access modifer

File: Calculator/Layout.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	 * Calculates the minimum size dimensions, given the components it contain
+	 */
+	void minimumLayoutSize() {
+	}
+
+	/**
+	 * Calculates the preferred size dimensions, given the components it //
+	 * contains
+	 */
+	void preferredLayoutSize() {
```
Comment (chirag-metacube) [inline Calculator/Layout.java:32]: Same comment as minimumLayoutSize

File: Calculator/Operations.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+public class Operations {
+
+	/**
+	 * Calculate according to operator
+	 * @param firstNum - first Number
+	 * @param secondNum - second Number
+	 * @param operator - operation to be performed
+	 * @return result for further calculation if any
+	 */
+	public int evaluate(int firstNum, int secondNum, char operator)
```
Comment (chirag-metacube) [inline Calculator/Operations.java:15]: For Calculator, Numbers should be in decimal

File: Matrix/Matrix.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	 * @param row
+	 *            - total number of rows
+	 * @param col
+	 *            - total number of columns
+	 */
+
+	public Matrix(int row, int col) {
+		this.noRows = row;
+		this.noCols = col;
+		data = new int[row][col];
```
Comment (chirag-metacube) [inline Matrix/Matrix.java:25]: Why just data? not this.data same as this.noRows and this.noCols

File: Matrix/Matrix.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		return transposeMatrix;
+	}
+
+	/**
+	 * perform matrix multiplication
+	 * 
+	 * @param secondMatrix
+	 * @return resultMatrix
+	 */
+
```
Comment (chirag-metacube) [inline Matrix/Matrix.java:75]: Extra line can be removed

File: Matrix/Matrix.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	public Matrix multiplication(Matrix secondMatrix) {
+		/*
+		 * number of column of first matrix is not equal to number of rows of
+		 * second matrix and return null
+		 */
+		if (this.noCols != secondMatrix.noRows) {
```
Comment (chirag-metacube) [inline Matrix/Matrix.java:81]: Nice

File: Matrix/Matrix.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+				sum = 0;
+				for (int k = 0; k < this.noCols; k++) {
+					/*
+					 * first matrix Elements at [i][k] multiply with second
+					 * matrix Elements at [k][j] and add the result into result
+					 * matrix
+					 */
+					sum += this.data[rowIndex][k]
+							* secondMatrix.data[k][columnIndex];
+					resultMatrix.addElements(rowIndex, columnIndex, sum);
```
Comment (chirag-metacube) [inline Matrix/Matrix.java:98]: resultMatrix.addElements(rowIndex, columnIndex, sum);
This line can be moved outside loop, as get the total sum of rows & column multiplication, when loop iteration ends.

File: Matrix/MatrixTest.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+import org.junit.Test;
+
+public class MatrixTest {
+
+	/*
+	 * happy test cases
+	 */
+	@Test
+	public void addElementsHappyTestCase() {
+		Matrix object = new Matrix(2, 2);
```
Comment (chirag-metacube) [inline Matrix/MatrixTest.java:13]: Matrix creation and adding elements is common in several tests, this can be moved in @before or can be done in a separate method

File: Matrix/MatrixTest.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+		secondMatrix.addElements(0, 1, 4);
+		secondMatrix.addElements(1, 0, 7);
+		secondMatrix.addElements(1, 1, 8);
+		secondMatrix.addElements(0, 0, 2);
+		secondMatrix.addElements(0, 1, 4);		
+		Matrix result = firstMatrix.multiplication(secondMatrix);
+		assertNull(result);
+	}
+	@Test(expected = ArrayIndexOutOfBoundsException.class)
+	public void addElementsBoundaryTestCase() {
```
Comment (chirag-metacube) [inline Matrix/MatrixTest.java:84]: No need to write this test case, As we haven't handled negative values in code.

**Review by chirag-metacube — APPROVED** [review]
